### DISCUSSION TITLE
Writing to characteristic 3 levels deep

### DISCUSSION THREAD

**Discussion by Matthijsj20** [hbldh/bleak#1478] [Q&A]
* bleak version: 0.20.2
* Python version: 3.11.2
* Operating System: RaspberryOS
* BlueZ version (`bluetoothctl -v`) in case of Linux: 5.66

### Description
I'm trying to use read the heart rate sensor of a wearable. The wearable has the following GATT structure. 
![IMG_2074](https://github.com/hbldh/bleak/assets/45590635/0daaf389-ad18-416a-b88f-ec5ab2652625)
As you can see the wearable contains a service called Heart Rate with a characteristic called Heart Rate Measurement. I succeeded in connecting to the wearable and reading  values from the heart rate measurement characteristic. Nevertheless, i noticed the the wearable only starts actually measuring heart rate when "True" is send to the Client Characteristic Configuration. I don't know how to send a message to the Client Characteristic Configuration. As indicated by the red lines on the left side the configuration is one level lower (3 lines) then the Measurement. When i try to send a message to the UUID of the Configuration (2902) it says it's not found and also scanning on services and characteristics does not show the Configuration endpoint. With the NRF connect App i managed to send a message to the configuration thus confirming that the wearable is not the problem.  

How do i go one level deeper so i can send TRUE to the Client Characteristic Configuration so the wearable actually starts measuring heart rate?

**Comment by dlech** [comment]
This level is called a "descriptor" rather than a "characteristic" (the top level is a "service"). The Client Characteristic Configuration descriptor is special though and we use `start_notify()` to set a callback and enable it at the same time.

https://bleak.readthedocs.io/en/latest/api/client.html#bleak.BleakClient.start_notify

**Reply by Matthijsj20** [reply]
Thank you for your answer. In my code i indeed use the start_notify() function to set a callback. I managed to find the descriptor and use the write_gatt_descriptor() function to try to write "True" to the descriptor. Nevertheless, i'm getting a not permitted error back which is weird because when i use the NRF connect app i can write to the descriptor and i don't get a not permitted error. 

I used functions of the bleak library to find out the handle of the descriptor corresponding to the heart rate measurement characteristic and i'm almost 100% sure that i'm writing the the right one. Any idea what i could be doing wrong?

I use the following 3 functions. The first function is the callback, the second function sets the callback function by calling the start_notify() function and the third function tries to write "True" to the descriptor.

```python
def callback(sender: BleakGATTCharacteristic, data: bytearray):
    print(f"{sender}::: {data}")

async def main(address):
    async with BleakClient(address) as client:
        await client.start_notify(MODEL_NBR_UUID, callback)

async def enable_hr(addres):
    async with BleakClient(address) as client:
        await client.write_gatt_descriptor(26,True) 
```

**Reply by dlech** [reply]
> In my code i indeed use the start_notify() function to set a callback. I managed to find the descriptor and use the write_gatt_descriptor() function to try to write "True" to the descriptor.

`write_gatt_descriptor()` is not needed in this case. `start_notify()` already does the for you.

**Reply by dlech** [reply]
> ```python
> async def main(address):
>     async with BleakClient(address) as client:
>         await client.start_notify(MODEL_NBR_UUID, callback)
> ```

You need something to keep the connection open here, like `await asyncio.Event().wait()`. Otherwise the connection disconnects when the async context manager exits.


```python
async def main(address):
    async with BleakClient(address) as client:
        await client.start_notify(MODEL_NBR_UUID, callback)
        # wait forever
        await asyncio.Event().wait()
```

**Reply by Matthijsj20** [reply]
Thank you again, the wait() function indeed fixed the problem and it is working right now. Thanks a lot!
